include/aws/sdkutils/endpoints/ruleset.h

```diff
@@ -0,0 +1,123 @@
+#define AWS_SDKUTILS_ENDPOINTS_RULESET_H
+
+#include <aws/common/byte_buf.h>
+#include <aws/sdkutils/sdkutils.h>
+
+struct aws_endpoints_ruleset;
+struct aws_endpoints_parameter;
+struct aws_hash_table;
+
+enum aws_endpoints_parameter_value_type { AWS_ENDPOINTS_PARAMETER_STRING, AWS_ENDPOINTS_PARAMETER_BOOLEAN };
+
+AWS_EXTERN_C_BEGIN
+
+/*
+******************************
+* Parameter
+******************************
+*/
+
+/*
+ * Value type of parameter.
+ */
+AWS_SDKUTILS_API enum aws_endpoints_parameter_value_type aws_endpoints_parameter_get_value_type(
+    const struct aws_endpoints_parameter *parameter);
+
+/*
+ * Parameter to built-in mapping.
+ */
+AWS_SDKUTILS_API const struct aws_string *aws_endpoints_parameter_get_built_in(
+    const struct aws_endpoints_parameter *parameter);
+
+/*
+ * Default string value. Returns AWS_OP_ERR if parameter is not a string.
+ */
+AWS_SDKUTILS_API int aws_endpoints_parameter_get_default_string(
+    const struct aws_endpoints_parameter *parameter,
+    struct aws_string **out_string);
+
+/*
+ * Default boolean value. Returns AWS_OP_ERR if parameter is not a boolean.
+ */
+AWS_SDKUTILS_API int aws_endpoints_parameter_get_default_boolean(
+    const struct aws_endpoints_parameter *parameter,
+    bool *out_bool);
+
+/*
+ * Whether parameter is required.
+ */
+AWS_SDKUTILS_API bool aws_endpoints_parameter_get_is_required(const struct aws_endpoints_parameter *parameter);
+
+/*
+ * Parameter documentation.
+ */
+AWS_SDKUTILS_API const struct aws_string *aws_endpoints_parameter_get_documentation(
+    const struct aws_endpoints_parameter *parameter);
+
+/*
+ * Whether parameter is deprecated.
+ */
+AWS_SDKUTILS_API bool aws_endpoints_parameters_get_is_deprecated(const struct aws_endpoints_parameter *parameter);
+
+/*
+ * Deprecation message. Null if parameter is not deprecated.
+ */
+AWS_SDKUTILS_API const struct aws_string *aws_endpoints_parameter_get_deprecated_message(
+    const struct aws_endpoints_parameter *parameter);
+
+/*
+ * Deprecated since. Null if parameter is not deprecated.
+ */
+AWS_SDKUTILS_API const struct aws_string *aws_endpoints_parameter_get_deprecated_since(
+    const struct aws_endpoints_parameter *parameter);
+
+/*
+******************************
+* Ruleset
+******************************
+*/
+
+/*
+ * Create new endpoint from a json string.
+ * In cases of failure NULL is returned and last error is set.
+ */
+AWS_SDKUTILS_API
+struct aws_endpoints_ruleset *aws_endpoints_ruleset_new_from_string(
+    struct aws_allocator *allocator,
+    const struct aws_byte_cursor *ruleset_cur);
+
+/*
+ * Increment ref count
+ */
+AWS_SDKUTILS_API void aws_endpoints_ruleset_acquire(struct aws_endpoints_ruleset *ruleset);
+
+/*
+ * Decrement ref count
+ */
+AWS_SDKUTILS_API void aws_endpoints_ruleset_release(struct aws_endpoints_ruleset *ruleset);
+
+/*
+ * Get ruleset parameters.
+ */
+AWS_SDKUTILS_API const struct aws_hash_table *aws_endpoints_ruleset_get_parameters(
```

Comment:
It will get bound out. We haven't bound out hash table before as far as i can tell. 
Since parameters are just informational structs, i was thinking of just doing a deep copy of the table and creating corresponding params in the binding, which would avoid having to do refcounting on params.

Reply:
Add a comment to this function's docs stating how the bindings should work.

I've [done that before](https://github.com/awslabs/aws-c-io/blob/442f9236db6f4c9b970fe0a08d9cda50bc444a81/include/aws/io/tls_channel_handler.h#L423)